### ISSUE TITLE
Provide a context that flows from parent targets to child ones

### ISSUE THREAD

**Issue by ygel** [adamralph/bullseye#287]
Hi, when a target chain executes it would be nice to have a notion of "context" created when dependency chain starts executing. This context object would then flow through to the outer tasks. This would help with resolve->download->unpack->process target chains.

**Comment by adamralph** [comment]
Hi @ygel, thanks for raising this. Can you please give an example of how you would like to use this context object?

**Comment by ygel** [comment]
Hello, the idea is two fold:
- to have a set of shared targets in a chain that accept a context to work on. Otherwise I'll have to effectively copy/paste same download/unpack calls to a global utility object.
- to have driver target in the dependency chain pass generated artifacts to its dependent target.

Since bullseye can run targets in parallel, using a global state objects would not be ideal.

Now, I tried to see if I can mock this up ... not a trivial task it turns out.

**Comment by adamralph** [comment]
@ygel naively, I'd say you can do that kind of thing by capturing local variables, e.g.:

```c#
Foo foo = default;
Bar bar = default;
Baz baz = default;

Target("foo", () => foo = GetFooIncludingDownloadingUnpackingAndWhateverElse())

Target("bar", DependsOn("foo"), () => bar = foo.CreateBar());

Target("baz", DependsOn("foo"), () => baz = foo.CreateBaz());

Target("bazz", DependsOn("bar", "baz"), () => PerformMagic(bar, baz));
```

Or is there more to it?

**Comment by ygel** [comment]
It's more along the lines of the following. 

    public class Foo { public async Task Build() { } }
    public class Bar { public async Task Build() { } }

    async Task BuildTaskSetup()
    {
        Target(nameof(Foo), DependsOn(nameof(Bar), () => new Foo().Build);
        Target(nameof(Bar), () => new Bar().Build);
    }

**Comment by adamralph** [comment]
> Having spelled this out I realize now that I can pass command line context to all tasks at creation time and they can pick out things they need, (which I already do, just not to the extent I should).

@ygel does that mean that you have solved your problem, and you do not need any changes to Bullseye?

> By the way: having #230 on nuget would be uber useful!

Sure, I'll push out a 3.0.0 alpha package soon. I'll ping you when it's available.

**Comment by ygel** [comment]
> @ygel does that mean that you have solved your problem, and you do not need any changes to Bullseye?

In progress of working it out, I'll update.

**Comment by ygel** [comment]
> Sure, I'll push out a 3.0.0 alpha package soon. I'll ping you when it's available.

Tested 3.0.0-beta.1, case-insensitive targets work.

**Comment by ygel** [comment]
See https://github.com/adamralph/bullseye/pull/325

**Comment by adamralph** [comment]
Closing, because #326 has been withdrawn.
